### PR TITLE
feat: add history for cost models and set view table for latest models

### PR DESCRIPTION
**Pull request by carlosvdr**
Reason for this PR. 

tl;dr: needed for TAP cost model checks

We currently have an issue where when a cost model just got updated  the indexer-service takes some time to update but in the meantime it errors out due to the model not matching, so we need to have a history in the db so that we can accept both, the new and the old one for a delimited amount of time (eg. ~2min)
This PR includes a breaking change:

- CostModels table no longer exists
- cost_models table is created, this will contain a history of all past cost models
- CostModels table will keep working as a view from cost_models this way other programs that read into CostModels will keep working as they are, this view will contain only the latest model of each deployment thus not breaking anything else

### CODE REVIEW

**Review by aasseman — CHANGES_REQUESTED** [review]

File: packages/indexer-agent/src/db/migrations/14-cost-models-history.ts [hunk 1]
```diff
@@ -17,6 +17,10 @@ export async function up({ context }: Context): Promise<void> {
   const tables = await queryInterface.showAllTables()
   logger.debug(`Checking if CostModelsHistory table exists`, { tables })
 
+  // CostModelsHistory: this table will store the history of cost models
+  // this is necessary since there could be a mismtach between the old table and the info the gateway has
+  // causing a failed request. Solution is to have a history and allow the "old" model for a limited time frame 2-3 mins
```
Comment (aasseman) [inline packages/indexer-agent/src/db/migrations/14-cost-models-history.ts:22]: You also need to give the context about tap-agent checking the receipt values

File: packages/indexer-common/src/indexer-management/resolvers/cost-models.ts [hunk 2]
```diff
@@ -80,14 +80,21 @@ export default {
...
+       "updatedAt"
+      FROM "CostModelsHistory" t1
+      JOIN
+      (
+          SELECT MAX(id)
+          FROM "CostModelsHistory"
+          ${deployments ? 'WHERE deployment IN (:deployments)' : ''}
+          GROUP BY deployment
+      ) t2
+        ON t1.id = t2.MAX;
```
Comment (aasseman) [inline packages/indexer-common/src/indexer-management/resolvers/cost-models.ts:97]: Can you also use this query in the view?

**Review by aasseman — CHANGES_REQUESTED** [review]

File: packages/indexer-agent/src/db/migrations/14-cost-models-history.ts [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+      updatedAt: {
+        type: DataTypes.DATE,
+        allowNull: false,
+      },
+    })
+    if (tables.includes('CostModels')) {
+      logger.debug(`Copying data from CostModels into CostModelsHistory`)
+      const copyTableSQL = `
+          INSERT INTO "CostModelsHistory" (id, deployment, model, variables, "createdAt", "updatedAt") 
+          SELECT id, deployment, model, variables, "createdAt", "updatedAt" FROM "CostModels";
```
Comment (aasseman) [inline packages/indexer-agent/src/db/migrations/14-cost-models-history.ts:79]: You shouldn't need to copy the ID, just let PG create new ones

File: packages/indexer-agent/src/db/migrations/14-cost-models-history.ts [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+       "updatedAt"
+      FROM "CostModelsHistory" t1
+      JOIN
+      (
+          SELECT MAX(id)
+          FROM "CostModelsHistory"
+          GROUP BY deployment
+      ) t2
+        ON t1.id = t2.MAX;
+      ;
```
Comment (aasseman) [inline packages/indexer-agent/src/db/migrations/14-cost-models-history.ts:106]: Extra `;`

File: packages/indexer-common/src/indexer-management/resolvers/cost-models.ts [hunk 5]
```diff
@@ -73,14 +75,38 @@ export default {
...
+      ;
```
Comment (aasseman) [inline packages/indexer-common/src/indexer-management/resolvers/cost-models.ts:98]: extra `;`

File: packages/indexer-common/src/indexer-management/resolvers/cost-models.ts [hunk 6]
```diff
@@ -121,22 +146,17 @@ export default {
     } catch (err) {
       throw new Error(`Invalid cost model or variables: ${err.message}`)
     }
-    const [model] = await models.CostModel.findOrBuild({
+    const [oldModel] = await models.CostModel.findOrBuild({
```
Comment (aasseman) [inline packages/indexer-common/src/indexer-management/resolvers/cost-models.ts:149]: Probably better not to build. Just try to find the latest model, and if there isn't one don't try to copy values over.

File: packages/indexer-common/src/indexer-management/resolvers/cost-models.ts [hunk 7]
```diff
@@ -121,22 +146,17 @@ export default {
...
       where: { deployment: update.deployment },
+      order: [['createdAt', 'DESC']],
```
Comment (aasseman) [inline packages/indexer-common/src/indexer-management/resolvers/cost-models.ts:151]: Order by ID instead. Last ID will always be the freshest one.

File: packages/indexer-common/src/indexer-management/resolvers/cost-models.ts [hunk 8]
```diff
@@ -121,22 +146,17 @@ export default {
...
-    // logger.info('Merged models', { now: model })
-    model.deployment = update.deployment || model.deployment
-    model.model = update.model || model.model
-
-    // Update the model variables (fall back to current value if unchanged)
-    let variables = update.variables || model.variables
+    const model = models.CostModel.build()
 
+    model.deployment = update.deployment || oldModel.deployment
+    model.model = update.model || oldModel.model
```
Comment (aasseman) [inline packages/indexer-common/src/indexer-management/resolvers/cost-models.ts:156]: I think it'd be cleaner to create the new entry in one go with by giving all the values in `build`, instead of creating then updating the row column by column.
Example from the [sequelize docs](https://sequelize.org/docs/v6/core-concepts/getters-setters-virtuals/#setters):

```js
const user = User.build({
  username: 'someone',
  password: 'NotSo§tr0ngP4$SW0RD!',
});
console.log(user.password); // '7cfc84b8ea898bb72462e78b4643cfccd77e9f05678ec2ce78754147ba947acc'
console.log(user.getDataValue('password')); // '7cfc84b8ea898bb72462e78b4643cfccd77e9f05678ec2ce78754147ba947acc'
```

File: packages/indexer-common/src/indexer-management/resolvers/cost-models.ts [hunk 9]
```diff
@@ -73,14 +75,38 @@ export default {
...
+    const costModels = await sequelize.query(query, {
+      replacements: { deployments: deployments ? deployments : [] },
+      mapToModel: true,
+      model: models.CostModel,
     })
     const definedDeployments = new Set(costModels.map((model) => model.deployment))
     const undefinedDeployments = deployments?.filter((d) => !definedDeployments.has(d))
     const globalModel = await models.CostModel.findOne({
       where: { deployment: COST_MODEL_GLOBAL },
+      order: [['createdAt', 'DESC']],
```
Comment (aasseman) [inline packages/indexer-common/src/indexer-management/resolvers/cost-models.ts:109]: Order by ID instead

File: packages/indexer-common/src/indexer-management/resolvers/cost-models.ts [hunk 10]
```diff
@@ -52,13 +52,15 @@ export default {
...
       where: { deployment },
+      order: [['createdAt', 'DESC']],
     })
     if (model) {
       return model.toGraphQL()
     }
 
     const globalModel = await models.CostModel.findOne({
       where: { deployment: COST_MODEL_GLOBAL },
+      order: [['createdAt', 'DESC']],
```
Comment (aasseman) [inline packages/indexer-common/src/indexer-management/resolvers/cost-models.ts:63]: Order by ID instead

File: packages/indexer-common/src/indexer-management/resolvers/cost-models.ts [hunk 11]
```diff
@@ -52,13 +52,15 @@ export default {
   ): Promise<object | null> => {
     const model = await models.CostModel.findOne({
       where: { deployment },
+      order: [['createdAt', 'DESC']],
```
Comment (aasseman) [inline packages/indexer-common/src/indexer-management/resolvers/cost-models.ts:55]: Order by ID instead

**Review by aasseman — CHANGES_REQUESTED** [review]

**Review by aasseman — COMMENTED** [review]
LGTM, please squash everything into a single commit so we can get the maintainers to review

**Review by aasseman — APPROVED** [review]

**Comment by dwerner** [comment]
looks good but let's get this in after we tag next release

**Review by dwerner — APPROVED** [review]
